packages/maskbook/src/plugins/ITO/UI/ClaimDialog.tsx

```diff
@@ -0,0 +1,357 @@
+    const chainIdValid = useChainIdValid()
+
+    const [ratio, setRatio] = useState<BigNumber>(
+        new BigNumber(payload.exchange_amounts[0 * 2]).dividedBy(new BigNumber(payload.exchange_amounts[0 * 2 + 1])),
+    )
+    const [claimToken, setClaimToken] = useState<EtherTokenDetailed | ERC20TokenDetailed>(payload.exchange_tokens[0])
+    const [claimAmount, setClaimAmount] = useState<BigNumber>(initAmount.multipliedBy(ratio))
+    const [inputAmountForUI, setInputAmountForUI] = useState(formatBalance(claimAmount, claimToken.decimals))
+
+    //#region select token
+    const [openSwapTokenDialog, setOpenSwapTokenDialog] = useState(false)
+    //#endregion
+
+    //#region balance
+    const { value: tokenBalance = '0', loading: tokenBalanceLoading } = useTokenBalance(
+        claimToken.type,
+        claimToken.address,
+    )
+    //#endregion
+
+    //#region remote controlled select provider dialog
+    const [, setSelectProviderDialogOpen] = useRemoteControlledDialog(WalletMessages.events.selectProviderDialogUpdated)
+    const onConnect = useCallback(() => {
+        setSelectProviderDialogOpen({
+            open: true,
+        })
+    }, [setSelectProviderDialogOpen])
+    //#endregion
+
+    //#region approve
+    const ITO_CONTRACT_ADDRESS = useConstant(ITO_CONSTANTS, 'ITO_CONTRACT_ADDRESS')
+    const [approveState, approveCallback] = useERC20TokenApproveCallback(
+        claimToken.type === EthereumTokenType.ERC20 ? claimToken.address : '',
+        claimAmount.toFixed(),
+        ITO_CONTRACT_ADDRESS,
+    )
+
+    const onApprove = useCallback(async () => {
+        if (approveState !== ApproveState.NOT_APPROVED) return
+        await approveCallback()
+    }, [approveState, approveCallback])
+    const onExactApprove = useCallback(async () => {
+        if (approveState !== ApproveState.NOT_APPROVED) return
+        await approveCallback(true)
+    }, [approveState, approveCallback])
+    const approveRequired =
+        (approveState === ApproveState.NOT_APPROVED || approveState === ApproveState.PENDING) &&
+        claimToken.type !== EthereumTokenType.Ether
+    //#endregion
+
+    //#region claim
+    const [claimState, claimCallback, resetClaimCallback] = useClaimCallback(
+        payload.pid,
+        payload.password,
+        claimAmount.toFixed(),
+        claimToken,
+    )
+    const onClaim = useCallback(async () => {
+        await claimCallback()
+        if (payload.token.type === EthereumTokenType.ERC20) {
+            await WalletRPC.addERC20Token(payload.token)
+            await WalletRPC.trustERC20Token(account, payload.token)
+        }
+    }, [account, payload, claimCallback])
+
+    const [_, setTransactionDialogOpen] = useRemoteControlledDialog(
+        WalletMessages.events.transactionDialogUpdated,
+        (ev) => {
+            if (ev.open) return
+            if (claimState.type !== TransactionStateType.CONFIRMED && claimState.type !== TransactionStateType.RECEIPT)
```

Comment:
```suggestion
            if (claimState.type !== TransactionStateType.CONFIRMED && claimState.type !== TransactionStateType.RECEIPT)
```
```suggestion
            if (claimState.type === TransactionStateType.TRANSACTION_HASH)
```

Reply:
I don't think so, when is still TRANSACTION_HASH, retryBuyInfo() and other hooks could not get the updated data.

Reply:
Never mind. That's a wrong fix.